File: app/CDoc/c/src/data/IndexTree.c
// Copyright 08-Dic-2021 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/IndexTree.h"
#include "dmc/DEFS.h"
#include "dmc/js.h"
#include "dmc/err.h"
#include "dmc/char/Achar.h"
#include "data/IndexTree/AIndexTree.h"

IndexTree *indexTree_new (char *id, Ochar *doc, AIndexTree *trees) {
  IndexTree *this = MALLOC(IndexTree);
  this->id = id;
  this->doc = doc;
  this->trees = trees;
  return this;
}

char *indexTree_to_js (IndexTree *this) {
  return js_wa(achar_new_from(
    js_ws(this->id),
    ochar_to_js(this->doc),
    aIndexTree_to_js(this->trees),
    NULL
  ));
}

IndexTree *indexTree_from_js (char *js) {
  return FAIL("Unimplemented function");
}

File: app/CDoc/js/src/data/IndexTree.js
// Copyright 18-Aug-2019 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/**
    Dpaths index tree.
**/
export default class IndexTree {

  /* .
  _rc_ IndexTree: from
    id: string
    isPath: boolean
    # If isPath is 'false', 'doc' is an empty string.
    doc: string
    # If isPath is 'true', 'tree' is 'null'.
    trees: !Array<!IndexTree>
  */

  /*--*/
  /**
      @param {string} id
      @param {boolean} isPath
      @param {string} doc If isPath is 'false', 'doc' is an empty string.
      @param {!Array<!IndexTree>} trees If isPath is 'true', 'tree' is 'null'.
  **/
  constructor (id, isPath, doc, trees) {

    /**
        @private
        @type {string}
    **/
    this._id = id;

    /**
        @private
        @type {boolean}
    **/
    this._isPath = isPath;

    /**
        @private
        @type {string}
    **/
    this._doc = doc;

    /**
        @private
        @type {!Array<!IndexTree>}
    **/
    this._trees = trees;

  }

  /**
      @return {string}
  **/
  get id () {
    return this._id;
  }

  /**
      @return {boolean}
  **/
  get isPath () {
    return this._isPath;
  }

  /**
      @return {string}
  **/
  get doc () {
    return this._doc;
  }

  /**
      @return {!Array<!IndexTree>}
  **/
  get trees () {
    return this._trees;
  }

  /**
      @param {!Array<?>} serial
      @return {!IndexTree}
  **/
  static fromJs (serial) {
    return new IndexTree(
      serial[0],
      serial[1],
      serial[2],
      serial[3].map(e =>
        IndexTree.fromJs(e)
      )
    );
  }
  /*--*/

}
